common/flatpak-transaction.c

```diff
@@ -2879,6 +2900,89 @@ flatpak_transaction_run (FlatpakTransaction *transaction,
   return FLATPAK_TRANSACTION_GET_CLASS (transaction)->run (transaction, cancellable, error);
 }
 
+static gboolean
+get_installation_dir_free_space (FlatpakInstallation  *installation,
+                                 guint64              *free_space,
+                                 GError              **error)
+{
+  g_autoptr (GFile) installation_dir = NULL;
+  g_autoptr (GFileInfo) info = NULL;
+
+  installation_dir = flatpak_installation_get_path (installation);
+
+  info = g_file_query_filesystem_info (installation_dir,
+                                       G_FILE_ATTRIBUTE_FILESYSTEM_FREE,
+                                       NULL,
+                                       error);
+  if (info == NULL)
+    return FALSE;
+
+  *free_space = g_file_info_get_attribute_uint64 (info, G_FILE_ATTRIBUTE_FILESYSTEM_FREE);
+  return TRUE;
+}
+
+static gboolean
+check_for_disk_space (FlatpakTransaction *self)
+{
+  FlatpakTransactionPrivate *priv = flatpak_transaction_get_instance_private (self);
+  g_autoptr (GFile) installation_dir = NULL;
+  g_autoptr (GError) error = NULL;
+  GList *l;
+  guint64 space_required = 0;
+  guint64 min_free_space = 0;
+  guint64 free_disk_space = 0;
+
+  g_debug ("Checking for free disk space for the current transaction");
+
+  for (l = priv->ops; l != NULL; l = l->next)
+    {
+      FlatpakTransactionOperation *op = l->data;
+
+      if (op->kind == FLATPAK_TRANSACTION_OPERATION_INSTALL)
+        space_required += flatpak_transaction_operation_get_installed_size (op);
+      else if (op->kind == FLATPAK_TRANSACTION_OPERATION_UPDATE)
+        {
+          /* Runtime updates typically are small hence keep their sizes out of
+           * of the disk space check.*/
+          if (g_str_has_prefix (op->ref, "runtime/"))
+            continue;
+
+          /* An app's worst possible case for update is that, all bits of the app
+           * are changed. Therefore, consider new installed size from remote metadata in
+           * order to update the app (because pruning happens at the end of transaction
+           * so we need this much available space anyway before updating it). */
+          space_required += flatpak_transaction_operation_get_installed_size (op);
+        }
+    }
+
+  installation_dir = flatpak_installation_get_path (priv->installation);
+  if (!flatpak_installation_get_min_free_space_bytes (priv->installation,
+                                                      &min_free_space,
+                                                      &error))
+    {
+      g_autofree gchar *installation_path = g_file_get_path (installation_dir);
+
+      g_warning ("Failed to query min-free-space for %s installation: %s",
```

Comment:
In what situations can this fail?

Reply:
On `fstat` failure or overflow while calculation min_free_space bytes : https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-repo.c#L2766-L2794